### PR TITLE
Refactor and move some invoke_simd tests from devel repo

### PR DESCRIPTION
**Pull request by fveselov**
First chunk of tests targeting invoke_simd feature from invoke-simd-tests development repo, moved with small changes.

### CODE REVIEW

**Review by kbobrovs — CHANGES_REQUESTED** [review]
@fveselov, thanks for doing this!

One general comment before we dive into the review is - please factor out all duplicated code - e.g. ESIMDSelector and createExceptionHandler are available from esimd_test_utils.hpp. I know not all tests follow that, but this would save us a lot of maintenance efforts.

File: SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
+// TODO: enable on Windows once driver is ready
+// REQUIRES: gpu && linux
+// UNSUPPORTED: cuda || hip
+//
+// Check that full compilation works:
+// RUN: not %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr %s -o %t.out 2>&1 | FileCheck %s
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp:6]: could you please please move negative non-executable tests into a separate subdir?

File: SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      } catch (std::exception &e) {
+        std::cout << "std::exception: " << e.what() << std::endl;
+      } catch (...) {
+        std::cout << "generic exception\n";
+      }
+    }
+  };
+}
+
+#ifndef INVOKE_SIMD
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp:139]: dead macro

File: SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    auto dev = q.get_device();
+    std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+    auto e = q.submit([&](handler &cgh) {
+      auto PA = bufa.get_access<access::mode::read>(cgh);
+      auto PB = bufb.get_access<access::mode::read>(cgh);
+      auto PC = bufc.get_access<access::mode::write>(cgh);
+
+      cgh.parallel_for<class Test>(
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp:176]: Please minimize the test. For negative test it is enough to use SYCL_EXTERNAL ESIMD functions as a starting point, w/o kernels, queues, exception handlers etc.
```
SYCL_EXTERNAL void foo() SYCL_ESIMD_FUNCTION {
  // test code
}
```

File: SYCL/InvokeSimd/ESIMD_to_unmarked_function_invoke_simd_new.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+ * From Konst:
+ *
+ * "Today we don’t support calling
+ * SYCL_EXTERNAL SPMD functions from ESIMD code, as semantics is not fully clear
+ * and will require specification.
+ * SYCL_EXTERNAL function is SPMD if it is not marked with SYCL_ESIMD_FUNCTION.
+ * W/o SYCL_EXTERNAL, a function w/o SYCL_ESIMD_FUNCTION is treated as ESIMD if
+ * called from within ESIMD code (callgraph is built). The same function can be
+ * called from SYCL, in which case there will be two copies created and compiled
+ * differently (via scalar and vector BEs)."
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/ESIMD_to_unmarked_function_invoke_simd_new.cpp:49]: Suggest to remove this. The test is not calling ESIMD->SPMD.
```suggestion
```

File: SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+  From Konst:
+
+  "Today we don’t support calling SYCL_EXTERNAL SPMD functions from ESIMD code,
+  as semantics is not fully clear and will require specification.
+  SYCL_EXTERNAL function is SPMD if it is not marked with SYCL_ESIMD_FUNCTION.
+  W/o SYCL_EXTERNAL, a function w/o SYCL_ESIMD_FUNCTION is treated as ESIMD if
+  called from within ESIMD code (callgraph is built). The same function can be
+  called from SYCL, in which case there will be two copies created and compiled
+  differently (via scalar and vector BEs)."
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp:25]: Suggest to remove this. There are lots of things ESIMD spec does not specify, and which are kind of assumed. 
```suggestion
This test case is based on the assumption that ESIMD functions can't call SPMD functions, as there is no mechanism for this described in ESIMD specification.
```

File: SYCL/InvokeSimd/call_vadd_1d.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
+// TODO: enable on Windows once driver is ready
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/call_vadd_1d.cpp:1]: I think this test is a duplicate of invoke_simd_smoke.cpp

File: SYCL/InvokeSimd/call_vadd_1d.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+// UNSUPPORTED: cuda || hip
+//
+// Check that full compilation works:
+// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr %s -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+// Now device-compile:
+// RUN: %clangxx -fsycl -Xclang -fsycl-allow-func-ptr -fsycl-device-only -S %s -o %t.ll
+// ... then run post-link on the compilation result, checking output IR:
+// ... some modules had to be split, '-ir-output-only' can't be used
+// RUN: sycl-post-link -lower-esimd -S %t.ll -o %t.xxx.ll
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/call_vadd_1d.cpp:12]: Such tests which invoke sycl-post-link should normally go to the test/tools/sycl-post-link directory in the llvm repo, and use .ll as input

File: SYCL/InvokeSimd/call_vadd_1d_loop.cpp [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+// Check that full compilation works:
+// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr %s -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+// Now device-compile:
+// RUN: %clangxx -fsycl -Xclang -fsycl-allow-func-ptr -fsycl-device-only -S %s -o %t.ll
+// ... then run post-link on the compilation result, checking output IR:
+// ... some modules had to be split, '-ir-output-only' can't be used
+// RUN: sycl-post-link -lower-esimd -S %t.ll -o %t.xxx.ll
+// ... concat modules before FileCheck
+// RUN: find . -iregex .*%basename_t.*.xxx.*.ll | xargs cat | FileCheck %s
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/call_vadd_1d_loop.cpp:14]: I'm not sure if this (find, xargs) will work on Windows. I don't see any precedents in the llvm-test-suite. Can the test goal be achieved w/o this?
@bader, does Windows CI support find and xargs?

File: SYCL/InvokeSimd/call_vadd_1d_loop.cpp [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            // uint32_t wi_id = i + sg.get_local_id();
+            // This simpler offset computation also works,
+            // but using the original one above for
+            // consistency.
+            // int wi_id = ndi.get_global_linear_id();
+
+            if constexpr (use_invoke_simd) {
+              float res = invoke_simd(sg, SIMD_CALLEE, uniform{A}, uniform{B},
+                                      uniform{C}, wi_id);
+              // CHECK: %{{[A-z0-9_.]*}} = call x86_regcallcc float @_Z33__regcall3____builtin_invoke_simd{{[A-z0-9_.]*}}(<16 x float> (float addrspace(4)*, float addrspace(4)*, float addrspace(4)*, <16 x i32>)* noundef nonnull @{{[A-z0-9_.]*}}spmd2simd{{[A-z0-9_.]*}}, float addrspace(4)* noundef %{{[A-z0-9_.]*}}, float addrspace(4)* noundef %{{[A-z0-9_.]*}}, float addrspace(4)* noundef %{{[A-z0-9_.]*}}, i32 noundef %{{[A-z0-9_.]*}}) [[ATTRS1:#[0-9]+]]
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/call_vadd_1d_loop.cpp:167]: E2E tests should not check IR. This is already done by unit tests. So please remove from here (some of it may go to llvm/test/tools/sycl-post-link or similar if you feel it adds coverage).

File: SYCL/InvokeSimd/call_vadd_1d_spill.cpp [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+constexpr bool use_invoke_simd = INVOKE_SIMD != 0;
+
+int main(void) {
+
+  float *A = new float[Size];
+  float *B = new float[Size];
+  float *C = new float[Size];
+  // temps
+  float *P = new float[Size];
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/call_vadd_1d_spill.cpp:126]: memory leaks?

File: SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp (hunk 3 above)
Comment (fveselov) [reply SYCL/InvokeSimd/ESIMD_to_SPMD_error_call_vadd_1d.cpp]: Moved error message checking tests to clang_run_error directory and minimized them as much as possible.

File: SYCL/InvokeSimd/call_vadd_1d.cpp (hunk 6 above)
Comment (fveselov) [reply SYCL/InvokeSimd/call_vadd_1d.cpp]: Agree, test removed.

**Comment by fveselov** [comment]
> @fveselov, thanks for doing this!
> 
> One general comment before we dive into the review is - please factor out all duplicated code - e.g. ESIMDSelector and createExceptionHandler are available from esimd_test_utils.hpp. I know not all tests follow that, but this would save us a lot of maintenance efforts.

@kbobrovs, duplicating code (ESIMDSelector and createExceptionHandler) removed and currently I referred to [esimd_test_utils.hpp](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/esimd_test_utils.hpp) using relative path, like:
`#include "../ESIMD/esimd_test_utils.hpp"`

Alternatively I can set clang flag to add ESIMD directory to include search paths.
I found that both approaches used in ESIMD tests: [dpas_test1.cpp](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/dpas/dpas_test1.cpp) vs [linear.cpp](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/linear/linear.cpp)

So I'm not sure on which way is preferred.

File: SYCL/InvokeSimd/clang_run_error/ESIMD_to_SPMD.cpp [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+#include <type_traits>
+
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+
+// 1024 / 16 = 64: There will be 64 iterations that process 16 elements each
+constexpr int Size = 1024;
+constexpr int VL = 16;
+// constexpr int VL = 8; // For some reason, this vector length is not
+// supported; 16 seems to be hardcoded somewere in Konst's code
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/clang_run_error/ESIMD_to_SPMD.cpp:36]: Hardcoding VL to 16 was removed quite a while ago, so please try with the new dpcpp builds - https://github.com/intel/llvm/releases/.

Also, please replace "Konst" with references to code and/or specs, as external readers don't have a clue about personalities.

File: SYCL/InvokeSimd/fig_4_13_nd_range_matrix_multiply_accessor_get_pointer.cpp [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+            [=](nd_item<2> ndi) [[intel::reqd_sub_group_size(VL)]] {
+              int row_index = ndi.get_global_id(0);
+              int column_index = ndi.get_global_id(1);
+
+              if constexpr (use_invoke_simd) {
+                // Currently, uniform{accessor::get_pointer()} returns incorrect
+                // results if the pointer returned is not explicitly cast to an
+                // appropriate data type (i.e., int *, etc.). "To me this looks
+                // suspicious, maybe a bug. Might be related to address space
+                // discrepancy." - Konst.
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/fig_4_13_nd_range_matrix_multiply_accessor_get_pointer.cpp:164]: I suggest to at least mark failing test cases with TODO. The best would be to extract them into separate XFAIL tests.
```suggestion
                // TODO FIXME: Currently, uniform{accessor::get_pointer()} returns incorrect
                // results if the pointer returned is not explicitly cast to an
                // appropriate data type (i.e., int *, etc.).
```

File: SYCL/InvokeSimd/invoke_simd_struct.cpp [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+ * and then scaled by y).
+ *
+ * Test case implementation notes:
+ * -------------------------------
+ * This test case has been modified from Konst's baseline
+ * test case, invoke_simd.cpp. It simply extends it by
+ * adding a struct pointer parameter to the invoke_simd
+ * functions which use it to scale the original vector
+ * addition.
+ */
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/invoke_simd_struct.cpp:31]:
```suggestion
 * This test case is a modified
 * https://github.com/intel/llvm/blob/sycl/sycl/test/invoke_simd/invoke_simd.cpp.
 *  It simply extends it by
 * adding a struct pointer parameter to the invoke_simd
 * functions which use it to scale the original vector
 * addition.
 */
```

Comment (kbobrovs) [reply SYCL/InvokeSimd/invoke_simd_struct.cpp:31]: One general note - please do not force-push, this makes it harder to see prior review comments/modifications.
If you want to get latest code from upstream, use `git checkout your_branch && git merge intel/intel`, then use simple `git push` when ready to push.

File: SYCL/InvokeSimd/fig_4_13_nd_range_matrix_multiply_accessor_get_pointer.cpp (hunk 12 above)
Comment (fveselov) [reply SYCL/InvokeSimd/fig_4_13_nd_range_matrix_multiply_accessor_get_pointer.cpp:164]: It looks like this comment is outdated. I looked at earlier revisions of this test and static_cast was removed at some point, but comment remained.

**Comment by fveselov** [comment]
It looks like [SYCL/InvokeSimd/call_vadd_1d_spill/call_vadd_1d_spill_VISALTO.cpp](https://github.com/intel/llvm-test-suite/pull/1182/files#diff-3647a193f6fdab7b228c27512afb5328bac759feb4014fe20e68cc1f24873246) pass on Release driver, but fail on Debug and ReleaseInternal. Jira ticket already created, but I have to remove XFAIL mark.

File: SYCL/InvokeSimd/invoke_simd_smoke.cpp [hunk 14]
```diff
@@ -97,7 +72,8 @@ template <bool use_func_directly> bool test() {
   constexpr unsigned Size = 1024;
   constexpr unsigned GroupSize = 4 * VL;
 
-  queue q(ESIMDSelector{}, createExceptionHandler());
+  auto GPUSelector = gpu_selector{};
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/invoke_simd_smoke.cpp:75]: why these changes to the existing tests?

File: SYCL/InvokeSimd/popcnt.cpp [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+
+// Tests invoke_simd support in the compiler/headers
+
+/* Test case purpose:
+ * -----------------
+ * This invoke_simd program has been modeled after
+ * https://gitlab.devtools.intel.com/dhochman/sginvokesimd_tests/-/blob/master/popcnt/popcnt.cpp
+ * which is not a working program, rather some pre-existing test case
+ * that seems to be testing simd_mask functionality and the popcount()
```
Comment (kbobrovs) [inline SYCL/InvokeSimd/popcnt.cpp:19]: What does it mean 'seems'? popcnt is orthogonal to invoke_simd. What aspect of invoke_simd implementation this test checks, which is not checked by other tests?

File: SYCL/InvokeSimd/invoke_simd_smoke.cpp (hunk 14 above)
Comment (fveselov) [reply SYCL/InvokeSimd/invoke_simd_smoke.cpp:75]: This way it is possible to avoid using ESIMDSelector and createExceptionHandler, that are described in SYCL/ESIMD/esimd_test_utils.hpp.

esimd_test_utils.hpp seems to be a bit overkill for these tests, as none of it is expected to run on host device, only on gpu. So I decided it would be reasonable to hardcode gpu selector.

**Comment by fveselov** [comment]
In order to reduce size of this pull request:
- Complex tests moved into separate PR: https://github.com/intel/llvm-test-suite/pull/1299
- Benchmark tests moved into separate PR: https://github.com/intel/llvm-test-suite/pull/1300
